pkg/solana/logpoller/loader.go

```diff
@@ -0,0 +1,258 @@
+}
+
+func NewEncodedLogCollector(
+	client RPCClient,
+	parser ProgramEventProcessor,
+	lggr logger.Logger,
+) *EncodedLogCollector {
+	c := &EncodedLogCollector{
+		client:       client,
+		parser:       parser,
+		chSlot:       make(chan uint64, 1),
+		chBlock:      make(chan uint64, 1),
+		chJobs:       make(chan Job, 1),
+		lggr:         lggr,
+		rpcTimeLimit: 1 * time.Second,
+	}
+
+	c.Service, c.engine = services.Config{
+		Name: "EncodedLogCollector",
+		NewSubServices: func(lggr logger.Logger) []services.Service {
+			c.workers = NewWorkerGroup(DefaultWorkerCount, lggr)
+
+			return []services.Service{c.workers}
+		},
+		Start: c.start,
+		Close: c.close,
+	}.NewServiceEngine(lggr)
+
+	return c
+}
+
+func (c *EncodedLogCollector) BackfillForAddress(ctx context.Context, address string, fromSlot uint64) error {
+	pubKey, err := solana.PublicKeyFromBase58(address)
+	if err != nil {
+		return err
+	}
+
+	sigs, err := c.client.GetSignaturesForAddressWithOpts(ctx, pubKey, &rpc.GetSignaturesForAddressOpts{
+		Commitment:     rpc.CommitmentFinalized,
+		MinContextSlot: &fromSlot,
+	})
+	if err != nil {
+		return err
+	}
+
+	slots := make(map[uint64]struct{})
+
+	for _, sig := range sigs {
+		_, ok := slots[sig.Slot]
+		if !ok {
+			if err := c.workers.Do(ctx, &getTransactionsFromBlockJob{
+				slotNumber: sig.Slot,
+				client:     c.client,
+				parser:     c.parser,
+				chJobs:     c.chJobs,
+			}); err != nil {
+				return err
+			}
+
+			slots[sig.Slot] = struct{}{}
+		}
+	}
+
+	return nil
+}
+
+func (c *EncodedLogCollector) start(ctx context.Context) error {
+	c.engine.Go(c.runSlotPolling)
+	c.engine.Go(c.runSlotProcessing)
+	c.engine.Go(c.runBlockProcessing)
+	c.engine.Go(c.runJobProcessing)
+
+	return nil
+}
+
+func (c *EncodedLogCollector) close() error {
+	return nil
+}
+
+func (c *EncodedLogCollector) runSlotPolling(ctx context.Context) {
+	for {
+		timer := time.NewTimer(DefaultNextSlotPollingInterval)
+
+		select {
+		case <-ctx.Done():
+			timer.Stop()
+
+			return
+		case <-timer.C:
+			ctxB, cancel := context.WithTimeout(ctx, c.rpcTimeLimit)
+
+			// not to be run as a job, but as a blocking call
+			result, err := c.client.GetLatestBlockhash(ctxB, rpc.CommitmentFinalized)
+			if err != nil {
+				c.lggr.Info("failed to get latest blockhash", "err", err)
+				cancel()
+
+				continue
+			}
+
+			cancel()
+
+			// if the slot is not higher than the highest slot, skip it
+			if c.lastSentSlot.Load() >= result.Context.Slot {
+				continue
+			}
+
+			c.lastSentSlot.Store(result.Context.Slot)
+			c.chSlot <- result.Context.Slot
+		}
+
+		timer.Stop()
+	}
+}
+
+func (c *EncodedLogCollector) runSlotProcessing(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case slot := <-c.chSlot:
+			if slot > c.highestSlot.Load() {
+				c.highestSlot.Store(slot)
+
+				if !c.loadingBlocks.Load() {
+					c.loadingBlocks.Store(true)
+
+					// run routine to load blocks in slot range
+					go func(start, end uint64) {
```

Comment:
I think instead of assuming that one may or may not run before the other, I can structure the channel listener and sender such that if the listener isn't listening (because its processing), the sender drops the slot. This will allow polling to continue and the range listener will still pull an effective range regardless of whether it misses a slot or not.